synapse/lib/stormtypes.py

```diff
@@ -1044,19 +1044,33 @@ class Query(StormType):
     '''
     A storm primitive representing an embedded query.
     '''
-    def __init__(self, text, opts, path=None):
+    def __init__(self, text, opts, runt, path=None):
 
         StormType.__init__(self, path=path)
 
         self.text = text
         self.opts = opts
+        self.runt = runt
 
         self.locls.update({
+            'exec': self._methQueryExec,
         })
 
     def __str__(self):
         return self.text
 
+    async def _methQueryExec(self):
+        query = await self.runt.getStormQuery(self.text)
+        subrunt = await self.runt.getScopeRuntime(query)
+
+        try:
+            async for item in query.run(subrunt, genr=s_ast.agen()):
+                pass  # pragma: no cover
+        except s_ast.StormReturn as e:
+            return e.item
+        finally:
+            await self.runt.propBackGlobals(subrunt)
```

Comment:
is it valid to run this if we hit an asyncio.cancelled situation?

Reply:
With how I've setup Query.exec(), it's effectively a function that you can pass around that inherits the runt of the runt it's created in. So if we can prop back globals in normal functions (which we currently can), it makes sense to prop back here (at least in my current thinking).

Reply:
so, if we're cancelled we'll still run this prior to raising.  We can prevent that with a bool to track if we've been cancelled or not, like we do in the view.stormstream implementation.